File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/BlocksRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.BlockedUser
import com.bubbler.android.data.model.User
import okhttp3.RequestBody.Companion.toRequestBody

/** List / block / unblock users. */
open class BlocksRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun getBlockedUsers(): List<BlockedUser> =
        apiClient.get(Endpoints.USER_ME_BLOCKS, token = tokenStore.requireAccessToken())

    open suspend fun blockUser(username: String): User =
        apiClient.request(
            path = Endpoints.userBlock(username),
            method = "POST",
            token = tokenStore.requireAccessToken(),
            // OkHttp requires a body for POST even when the route takes none.
            body = ByteArray(0).toRequestBody(null),
        )

    open suspend fun unblockUser(username: String): User =
        apiClient.request(
            path = Endpoints.userBlock(username),
            method = "DELETE",
            token = tokenStore.requireAccessToken(),
        )
}

File: bubbler_app/lib/data/repositories/blocks_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/blocked_user.dart';
import '../models/user.dart';

/// List / block / unblock users.
class BlocksRepository {
  BlocksRepository(this._client);

  final ApiClient _client;

  /// Blocked users for the signed-in account (`GET /user/me/blocks`).
  Future<List<BlockedUser>> getBlockedUsers() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeBlocks,
    );
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data
        .map((e) => BlockedUser.fromJson(_asJsonMap(e)))
        .toList();
  }

  /// Blocks [username] (`POST /user/me/blocks/{username}`).
  Future<User> blockUser(String username) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeBlock(username),
      method: 'POST',
    );
    return User.fromJson(_asJsonMap(data));
  }

  /// Unblocks [username] (`DELETE /user/me/blocks/{username}`).
  Future<User> unblockUser(String username) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeBlock(username),
      method: 'DELETE',
    );
    return User.fromJson(_asJsonMap(data));
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}
